### PR TITLE
RavenDB-12377 Backup database record in backup/restore

### PR DESCRIPTION
**Pull request by efratshenhar**

### CODE REVIEW

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 1]
```diff
@@ -118,37 +123,844 @@ public DatabaseRecordActions(BlittableJsonTextWriter writer)
...
+                            _writer.WriteComma();
+                            _writer.WritePropertyName(nameof(databaseRecord.SinkPullReplications));
+                            WriteSinkPullReplications(databaseRecord.SinkPullReplications);
+                        }
+
+                        break;
+                }
+            }
+
+            private void WriteHubPullReplications(List<PullReplicationDefinition> hubPullReplications)
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs:245]: Can't we do something like `context.Write(_writer, pullReplication.ToJson())`, if possible would be much cleaner.

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs [hunk 2]
```diff
@@ -23,4 +23,25 @@ public enum DatabaseItemType
         Counters = 1 << 13,
         Attachments = 1 << 14
     }
+
+    [Flags]
+    public enum DatabaseRecordItemType
+    {
+        NotSet = 0,
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: Let's rename this to `None`

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 3]
```diff
@@ -103,6 +103,11 @@ public async Task PostExport()
                     {
                         var blittableJson = await context.ParseToMemoryAsync(stream, "DownloadOptions", BlittableJsonDocumentBuilder.UsageMode.None, buffer);
                         options = JsonDeserializationServer.DatabaseSmugglerOptions(blittableJson);
+                        if (((options.OperateOnTypes & DatabaseItemType.DatabaseRecord) != 0)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs]: Let's move this to `ApplyBackwardCompatibility` method

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 4]
```diff
@@ -669,6 +674,11 @@ public async Task PostImportAsync()
                                         }
 
                                         options = JsonDeserializationServer.DatabaseSmugglerOptions(blittableJson);
+                                        if (((options.OperateOnTypes & DatabaseItemType.DatabaseRecord) != 0)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs]: Let's move that to `ApplyBackwardCompatibility` method

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 5]
```diff
@@ -118,37 +123,844 @@ public DatabaseRecordActions(BlittableJsonTextWriter writer)
...
+            public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressBase.DatabaseRecordProgress progress, AuthorizationStatus authorizationStatus, DatabaseRecordItemType databaseRecordItemType)
             {
                 _writer.WritePropertyName(nameof(databaseRecord.DatabaseName));
                 _writer.WriteString(databaseRecord.DatabaseName);
                 _writer.WriteComma();
 
                 _writer.WritePropertyName(nameof(databaseRecord.Encrypted));
                 _writer.WriteBool(databaseRecord.Encrypted);
+
+                if ((databaseRecordItemType & DatabaseRecordItemType.ConflictSolverConfig) != 0)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: Let's use `.Contain()` extension method

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Smuggler/Documents/StreamDestination.cs:245]: We can use the `ToJson`, yes. This is not a hot-path. On other items too, if applicable.

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 6]
```diff
@@ -118,37 +123,844 @@ public DatabaseRecordActions(BlittableJsonTextWriter writer)
...
+            {
+                if (settings == null)
+                {
+                    _writer.WriteNull();
+                    return;
+                }
+
+                var serverWideKeys = DatabaseHelper.GetServerWideOnlyConfigurationKeys();
+
+                string[] doNotBackUp = new string[]
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: should be static hashset (case insensitive)

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 7]
```diff
@@ -118,37 +123,844 @@ public DatabaseRecordActions(BlittableJsonTextWriter writer)
...
+                {
+                    "DataDir", "Storage.TempPath", "Indexing.TempPath", "License", "RunInMemory"
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: use RavenConfiguration.GetKey to get those values

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 8]
```diff
@@ -118,37 +123,844 @@ public DatabaseRecordActions(BlittableJsonTextWriter writer)
...
+
+            private void WriteSettings(Dictionary<string, string> settings)
+            {
+                if (settings == null)
+                {
+                    _writer.WriteNull();
+                    return;
+                }
+
+                var serverWideKeys = DatabaseHelper.GetServerWideOnlyConfigurationKeys();
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: can also be static and needs to be case insensitive, can use hashset here

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 9]
```diff
@@ -385,6 +314,17 @@ private void WriteSorters(Dictionary<string, SorterDefinition> sorters)
                 _writer.WriteEndObject();
             }
 
+            private static readonly HashSet<string> DoNotBackUp = new HashSet<string>
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: We need to use StringComparer.OridinalIgnoreCase here

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 10]
```diff
@@ -385,6 +314,17 @@ private void WriteSorters(Dictionary<string, SorterDefinition> sorters)
...
+            {
+                RavenConfiguration.GetKey(x => x.Core.DataDirectory),
+                RavenConfiguration.GetKey(x => x.Storage.TempPath),
+                RavenConfiguration.GetKey(x => x.Indexing.TempPath),
+                RavenConfiguration.GetKey(x => x.Licensing.License),
+                RavenConfiguration.GetKey(x => x.Core.RunInMemory)
+            };
+
+            private static readonly HashSet<string> ServerWideKeys = DatabaseHelper.GetServerWideOnlyConfigurationKeys().ToHashSet();
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: We need to use StringComparer.OridinalIgnoreCase here

File: src/Raven.Server/Smuggler/Documents/StreamSource.cs [hunk 11]
```diff
@@ -269,30 +269,23 @@ public DatabaseRecord GetDatabaseRecord()
...
                         {
-                            var hub = JsonDeserializationCluster.PullReplicationDefinition(pullReplicationDefinition);
+                            var hub = JsonDeserializationCluster.PullReplicationDefinition(pullReplication);
                             databaseRecord.HubPullReplications.Add(hub);
                         }
                         catch (Exception e)
                         {
                             if (_log.IsInfoEnabled)
-                                _log.Info("Wasn't able to import hub pull replication configuration from smuggler file. Skipping.", e);
+                                _log.Info($"Wasn't able to import the pull Replication {pullReplication} configuration from smuggler file. Skipping.", e);
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamSource.cs]: typo: `Pull Replication`

Comment (ppekrol) [reply src/Raven.Server/Smuggler/Documents/StreamSource.cs]: What is inside the `pull replication` json. Any confidential data? If yes, then we probably do not want to write it to the log, just basic info.

**Comment by ppekrol** [comment]
Please add interversion test (talk to @maximburyak) checking if we can smuggler from 4.1 to 4.2 and from 4.2 to 4.1 and if backward compatibility works.

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 12]
```diff
@@ -108,90 +114,672 @@ public IIndexActions Indexes()
...
+                        _writer.WritePropertyName(config.Key);
+                        _writer.WriteString(config.Value);
+                        _writer.WriteEndObject();
+                    }
                 }
+                _writer.WriteEndArray();
+            }
 
-                if (clientConfiguration.PrettifyGeneratedLinqExpressions.HasValue)
+            private void WriteSqlEtls(List<SqlEtlConfiguration> sqlEtlConfiguration)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs:347]: Let's implement `ToJson` in `SqlEtlConfiguration

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 13]
```diff
@@ -108,90 +114,672 @@ public IIndexActions Indexes()
...
+                    _writer.WritePropertyName(nameof(etl.SqlTables));
+                    WriteSqlTables(etl.SqlTables);
+                    _writer.WriteEndObject();
                 }
 
-                if (clientConfiguration.ReadBalanceBehavior.HasValue)
+                _writer.WriteEndArray();
+            }
+
+            private void WriteSqlTables(List<SqlEtlTable> SqlTables)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: Let's implement `ToJson`

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 14]
```diff
@@ -108,90 +114,672 @@ public IIndexActions Indexes()
...
+                    if (first == false)
+                        _writer.WriteComma();
+                    first = false;
+                    _context.Write(_writer, sqlTable.ToJson());
+                }
+
+                _writer.WriteEndArray();
+            }
+
+            private void WriteRavenEtls(List<RavenEtlConfiguration> ravenEtlConfiguration)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs:368]: Let's implement `ToJson`

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 15]
```diff
@@ -108,90 +114,672 @@ public IIndexActions Indexes()
...
+                {
+                    if (first == false)
+                        _writer.WriteComma();
+                    first = false;
+                    WriteBackupConfiguration(backup);
                 }
+                _writer.WriteEndArray();
+            }
+
+            private void WriteBackupConfiguration(PeriodicBackupConfiguration backup)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: Let's implement `ToJson`

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTask.cs [hunk 16]
```diff
@@ -523,14 +523,26 @@ private List<string> GetFilesForRestore(string backupLocation)
                 },
                 onDatabaseRecordAction: smugglerDatabaseRecord =>
                 {
-                    // need to enable revisions before import
-                    database.DocumentsStorage.RevisionsStorage.InitializeFromDatabaseRecord(smugglerDatabaseRecord);
-
+                    databaseRecord.ConflictSolverConfig = smugglerDatabaseRecord.ConflictSolverConfig;
+                    foreach (var setting in smugglerDatabaseRecord.Settings)
+                    {
+                        databaseRecord.Settings[setting.Key] = setting.Value;
```
Comment (ayende) [inline src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTask.cs:529]: Aren't you overwriting the `DataDir` here?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 17]
```diff
@@ -445,11 +448,124 @@ public DatabaseRecordActions(DocumentDatabase database, Logger log)
...
+                }
+
+                if (currentDatabaseRecord?.SqlEtls.Count == 0 &&
+                    databaseRecord?.SqlEtls.Count > 0)
+                {
+                    if (_log.IsInfoEnabled)
+                        _log.Info("Configuring sql etls configuration from smuggler");
+                    foreach (var etl in databaseRecord.SqlEtls)
+                    {
+                        tasks.Add(_database.ServerStore.SendToLeaderAsync(new AddSqlEtlCommand(etl, _database.Name)));
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:564]: Why are you calling the commands here and not modifying the database record directly?

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 18]
```diff
@@ -108,246 +114,396 @@ public IIndexActions Indexes()
...
                 {
                     _writer.WriteComma();
-                    _writer.WritePropertyName(nameof(clientConfiguration.ReadBalanceBehavior));
-                    _writer.WriteString(clientConfiguration.ReadBalanceBehavior.Value.ToString());
+                    _writer.WritePropertyName(nameof(databaseRecord.Sorters));
+                    WriteSorters(databaseRecord.Sorters);
                 }
 
-                _writer.WriteEndObject();
+                switch (authorizationStatus)
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs:180]: Why do we have a check here?

File: src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTask.cs (hunk 16 above)
Comment (efratshenhar) [reply src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTask.cs:529]: No, I am not saving the DataDir in the backup

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs (hunk 18 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/StreamDestination.cs:180]: We are writing here certificates, password etc.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 17 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:564]: I need to get new information for the command. like, TaskId, mentor node etc.

**Pull request merged**